bpf/lib/conntrack.h

```diff
@@ -466,6 +467,26 @@ static __always_inline void ipv4_ct_tuple_reverse(struct ipv4_ct_tuple *tuple)
 	ct_flip_tuple_dir4(tuple);
 }
 
+static __always_inline int ipv4_ct_extract_l4_ports(struct __ctx_buff *ctx,
+						    int off,
+						    struct ipv4_ct_tuple *tuple)
+{
+#if defined IPV4_FRAGMENTS
+	void *data, *data_end;
+	struct iphdr *ip4;
+
+	// TODO: Can we optimise this out?
+	if (!revalidate_data(ctx, &data, &data_end, &ip4))
+		return DROP_INVALID;
```

Comment:
Why is it needed? Wasn't the IPv4 header already validated to be able to load the sport+dport? And `off` already points to the IPv4 header, no?

Reply:
This is called from `ct_lookup4()`, and bpf_lxc.c has the following warning just before calling that function from `handle_ipv4_from_lxc()`: 
``` c
	/* WARNING: ip4 offset check invalidated, revalidate before use */
```

Reply:
Ah, I see. This check wasn't needed before because `ctx_load_bytes` has it's own length check (even in XDP case).

I think the TODO can be removed (unless you had an idea in mind to optimize it out?).

Reply:
Not really, not without taking L4 ports extraction out of `ct_lookup4()`, but that would probably require quite some amount of refactoring. I'll drop the TODO.